### ISSUE TITLE
RangeError: Format string contains an unescaped latin alphabet character `z`

### ISSUE THREAD

**Issue by aaemnnosttv** [date-fns/date-fns#2088]
I'm getting this `RangeError` when trying to parse a date string like this `2020-11-09 6:05pm GMT` using `z` for the timezone as noted in the unicode [date symbol field table](https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table).

Similar to https://github.com/date-fns/date-fns/issues/1473 except this is about parsing rather than formatting.

I realize that the docs for [parse](https://date-fns.org/v2.16.1/docs/parse) don't list `z` as a pattern but using `x` or `X` or `xx` etc doesn't work as well. As an aside, the docs seem to imply that the functions support all of the patterns in the unicode standard – plus additions, but that doesn't seem to be the case.

I created a sandbox to demonstrate the specific error here: https://codesandbox.io/s/date-fns-parsing-issue-qkyht

Is this format not possible to parse?

Thanks ahead of time for the help!

**Comment by scscgit** [comment]
This issue seems to be quite important, did anyone at least find any proper workaround to be able to translate tie `z`-formatted timezones into offsets, or any way to use them in `date-fns-tz` instead of IANA `timeZone`? As one use-case, users may have to consume Java API that formats date with `z` using `java.text.SimpleDateFormat`. I wasn't able to find a way to convert the Dates this way using `toDate` or `zonedTimeToUtc` either. I'd suggest to update the documentation by including a `parse` scenario at https://date-fns.org/v2.16.1/docs/Time-Zones. Note that `format` properly recognizes the `z` without any issue, so it'd be preferable to list all similar missing features on a page like README.

**Comment by vadimpopa** [comment]
I can't find either how to parse '7/22/2021 5:41:26 vorm. OESZ' with :

```
import de from 'date-fns/locale/de'

parse('7/22/2021 5:41:26 vorm. OESZ' ,'M/d/yyyy h:mm:ss a z', new Date(), {
 locale: de,
});
```

Anyone succeed ?

**Comment by fturmel** [comment]
@vadimpopa This is a tricky one, and I don't think there's an easy solution here. I'll try my best to help.

First off, the error message you are getting is because the `parse` function doesn't support the `z` token, so it is asking you to escape your non-token characters explicitly with single quotes. The `x` and `X` tokens are supported, which can handle the string `Z` and timezone offsets. https://date-fns.org/docs/parse

Secondly, `OESZ` appears to be the German translation of `EEST` (Eastern European Summer Time), so that's quite a hurdle that your identifier is localized.

One possible approach could be to map these timezone abbreviations to their known offsets and transform the input string into something that `date-fns` can understand.

```js
import { parse } from "date-fns";
import { de } from "date-fns/locale";

const input = "7/22/2021 5:41:26 vorm. OESZ";

const offsets = [
  ["MESZ", "+0200"],
  ["OESZ", "+0300"]
];

const inputWithOffset = offsets.reduce(
  (acc, [abr, offset]) => acc.replace(abr, offset),
  input
);

const res = parse(inputWithOffset, "M/d/yyyy h:mm:ss a xx", new Date(), {
  locale: de
});

console.log(res.toISOString()); // 2021-07-22T02:41:26.000Z
```

The `date-fns-tz` companion library supports IANA timezones, and would allow you to use `Europe/Bucharest` for example. But I don't think it will know what to do with `OESZ`. From the quick test I ran, it thinks the input is in UTC because the identifier ends with `Z`. Also, you would still need to use the `parse` function somehow to deal with the localized time of day.

**Comment by vadimpopa** [comment]
@fturmel yes I understood that, after I looked through code and thought a little about it. I've found a workaround that won't need parsing...worst case who (from our clients) will want such a format that will need parsing, will have to use offset too in format, so before parsing will remove `z`, since offset can provide the timezone. I can't use mapping, as you suggested, as we have many languages to support, and mapping is a pain - but yeah mapping idea came first in my mind too, and probably is the only solution, generally.

**Comment by fturmel** [comment]
@vadimpopa Glad you figured out an approach, and yes I imagine this would get very unwieldy for several languages.

Little bonus code snippet update that will scale much better if a larger list of offsets is needed.

```js
import { parse } from "date-fns";
import { de } from "date-fns/locale";

const input = "7/22/2021 5:41:26 vorm. OESZ";

const offsets = {
  MESZ: "+0200",
  OESZ: "+0300"
};

const abrv = input.match(/ ([A-Z]+)$/)[1];
const offset = offsets[abrv];
const inputWithOffset = input.replace(abrv, offset);

const res = parse(inputWithOffset, "M/d/yyyy h:mm:ss a xx", new Date(), {
  locale: de
});

console.log(res.toISOString()); // 2021-07-22T02:41:26.000Z
```

And a few resources:
https://www.timeanddate.com/time/zones/
https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
https://www.iana.org/time-zones

**Comment by fturmel** [comment]
@aaemnnosttv really late, but I wanted to try to reply to your original issue as well.

The missing `O` and `Z` tokens are supposed to understand `GMT`, and should assume an offset of 0 if it is missing. I believe this could be implemented without problems, I will try to work on a PR.

Supporting `z` would mean bundling all abbreviated timezones in the project, so that's a bit of baggage to bring in for a non-essential token. Maybe there is a sensible approach to implementing this without penalizing bundle sizes for everyone. There is also the question of handling standard vs daylight saving time and applying the correct offset in context. Something to think about.

> As an aside, the docs seem to imply that the functions support all of the patterns in the unicode standard [...]

I don't think there is such a claim, but I suppose it could be explicitly said that not all tokens from the spec are implemented.

For now, here are some ideas to parse your example. I don't believe it can be done without transforming the input though, but it will be possible if the `O` or `Z` tokens get implemented and merged.

```js
import { parse } from "date-fns";

const input = "2020-11-09 6:05pm GMT";

// escape GMT in format, append Z
let dateString = input + "Z";
let dateFormat = "yyyy-MM-dd h:mma 'GMT'X";
let res = parse(dateString, dateFormat, new Date());
console.log(res.toISOString()); // 2020-11-09T18:05:00.000Z

// escape GMT in format, append 0 offset
dateString = input + "+00";
dateFormat = "yyyy-MM-dd h:mma 'GMT'x";
res = parse(dateString, dateFormat, new Date());
console.log(res.toISOString()); // 2020-11-09T18:05:00.000Z

// replace GMT with Z
dateString = input.replace("GMT", "Z");
dateFormat = "yyyy-MM-dd h:mma X";
res = parse(dateString, dateFormat, new Date());
console.log(res.toISOString()); // 2020-11-09T18:05:00.000Z
```

**Comment by fturmel** [comment]
@scscgit I like the idea of a page to compare token compatibility between the parse and format functions, making note of it.

> format properly recognizes the z without any issue 

Actually, it looks like `z` in `format` was implemented to behave like `O`... so that's going to be a problem for interop and a breaking change if it gets fixed.

```js
import { format } from "date-fns";

const date = new Date(2021, 0, 1);
console.log(format(date, "z")); // GMT-5
console.log(format(date, "zzzz")); // GMT-05:00
```

https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table

<img width="682" alt="Screen Shot 2022-01-26 at 3 26 57 PM" src="https://user-images.githubusercontent.com/4009209/151241563-497f5856-1687-4b2c-a4a7-2c568daf27ea.png">

**Comment by margaretjoanmiller** [comment]
Is there any effort to support z in parse? I'm dealing with several dates from external sources with the GMT style abbreviations?
